xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/DBListClass.java

```diff
@@ -282,8 +285,16 @@ public String getQuery(XWikiContext context)
                     select.append("doc." + idField);
                 } else {
                     select.append("idprop.value");
-                    fromStatements.add("StringProperty as idprop");
                     whereStatements.add("obj.id=idprop.id.id and idprop.id.name='" + idField + "'");
+                    // Get the from statements according to the type of property
+                    try {
+                        PropertyClass pc = (PropertyClass) context.getWiki().getDocument(classname, context).getXClass()
+                                .get(idField);
+                        String classType = (pc == null) ? "StringProperty" : pc.newProperty().getClass().getName();
+                        fromStatements.add(classType + " as idprop");
+                    } catch (XWikiException e) {
+                        e.printStackTrace();
```

Comment:
Actually a better solution would be to have a try/catch at the level of getQuery() with a more informative message like "Failed to compute query for DBList ....". You'd need to refactor a bit the code since it's already way too much nested.